File: src/policy/filebased.go
package policy

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"strings"

	"github.com/fatih/color"
	glob "github.com/gobwas/glob"
	yaml "gopkg.in/yaml.v2"
)

type PolicyConf struct {
	DiffToDefault    bool                         `yaml:"diff_to_default"`
	WhitelistPaths   map[PathOps][]PatternMatcher `yaml:"whitelist_paths"`
	ExecAllowance    int                          `yaml:"exec_allowance"`
	ForkAllowance    int                          `yaml:"fork_allowance"`
	MaxChildProcs    int                          `yaml:"max_child_procs"`
	ExtraEnvs        []string                     `yaml:"extra_envs"`
	PreservedEnvKeys []string                     `yaml:"preserved_env_keys"`
	TracedSyscalls   []string                     `yaml:"traced_syscalls"`
	AllowedSyscalls  []string                     `yaml:"allowed_syscalls"`
}

func (o *PathOps) UnmarshalYAML(unmarshal func(interface{}) error) error {
	var raw string
	if err := unmarshal(&raw); err != nil {
		return err
	}
	var ok bool
	if *o, ok = pathOpsNameMap[raw]; !ok {
		return fmt.Errorf("invalid path operation name: %s", raw)
	}
	return nil
}

func (p *PatternMatcher) UnmarshalYAML(unmarshal func(interface{}) error) error {
	var raw string
	var err error
	if err = unmarshal(&raw); err != nil {
		return err
	}
	var g glob.Glob
	if g, err = glob.Compile(raw); err != nil {
		return err
	}
	*p = PatternMatcher{g}
	return nil
}

type FileBasedPolicy struct {
	FileName string
	conf     PolicyConf
}

func (p FileBasedPolicy) CheckPathOp(path string, op PathOps, mode int) bool {
	for _, matcher := range p.conf.WhitelistPaths[op] {
		if matcher.Match(path) {
			return true
		}
	}
	return false
}

func (p FileBasedPolicy) GetExecAllowance() int {
	return p.conf.ExecAllowance
}

func (p FileBasedPolicy) GetForkAllowance() int {
	return p.conf.ForkAllowance
}

func (p FileBasedPolicy) GetMaxChildProcs() int {
	return p.conf.MaxChildProcs
}

func (p FileBasedPolicy) GetExtraEnvs() []string {
	return p.conf.ExtraEnvs
}

func (p FileBasedPolicy) GetPreservedEnvKeys() []string {
	return p.conf.PreservedEnvKeys
}

func (p FileBasedPolicy) GetTracedSyscalls() []string {
	return p.conf.TracedSyscalls
}

func (p FileBasedPolicy) GetAllowedSyscalls() []string {
	return p.conf.AllowedSyscalls
}

func GeneratePolicyFromYAML(policyFile string) (FileBasedPolicy, error) {

	var conf *PolicyConf = new(PolicyConf)
	l := log.New(os.Stderr, "", 0)

	if policyFile == "[default]" {
		*conf = defaultConf
	} else {
		yamlData, err := ioutil.ReadFile(policyFile)
		if err != nil {
			color.Set(color.FgYellow)
			l.Println("Cannot read the policy file. Falling back to default policy...")
			color.Unset()
			policyFile = "[default]"
			*conf = defaultConf
		} else {
			// Update default conf with custom config.
			err = yaml.Unmarshal(yamlData, conf)
			if err != nil {
				return FileBasedPolicy{}, err
			}
			// We use fixed set of traced syscalls.
			conf.TracedSyscalls = defaultConf.TracedSyscalls
		}
	}

	// Merge if the policy only states the additional differences of environ/syscalls to the default.
	// The resource limit counts are not merged but just used as-is.
	if conf.DiffToDefault {
		mergedExtraEnvs := make(map[string]string)
		mergedPreservedEnvKeys := make(map[string]bool)
		mergedAllowedSyscalls := make(map[string]bool)

		for _, e := range defaultConf.ExtraEnvs {
			split := strings.SplitN(e, "=", 2)
			mergedExtraEnvs[split[0]] = split[1]
		}
		for _, e := range conf.ExtraEnvs {
			split := strings.SplitN(e, "=", 2)
			mergedExtraEnvs[split[0]] = split[1]
		}
		conf.ExtraEnvs = nil
		for k, v := range mergedExtraEnvs {
			conf.ExtraEnvs = append(conf.ExtraEnvs, fmt.Sprintf("%s=%s", k, v))
		}

		for _, k := range defaultConf.PreservedEnvKeys {
			mergedPreservedEnvKeys[k] = true
		}
		for _, k := range conf.PreservedEnvKeys {
			mergedPreservedEnvKeys[k] = true
		}
		conf.PreservedEnvKeys = nil
		for k, _ := range mergedPreservedEnvKeys {
			conf.PreservedEnvKeys = append(conf.PreservedEnvKeys, k)
		}

		for _, s := range defaultConf.AllowedSyscalls {
			mergedAllowedSyscalls[s] = true
		}
		for _, s := range conf.AllowedSyscalls {
			mergedAllowedSyscalls[s] = true
		}
		conf.AllowedSyscalls = []string{}
		for s, _ := range mergedAllowedSyscalls {
			conf.AllowedSyscalls = append(conf.AllowedSyscalls, s)
		}
	}

	return FileBasedPolicy{policyFile, *conf}, nil
}

File: src/policy/filebased.rs
use crate::policy::traits::{SandboxPolicy, PathOps, PolicyConf, PolicyError};

use std::{fs::File, collections::{HashMap, HashSet}};

pub struct FileBasedPolicy {
    pub file_name: String,
    conf: PolicyConf,
}

impl FileBasedPolicy {
    pub fn generate_from_yaml(policy_file: &Option<String>) -> Result<FileBasedPolicy, PolicyError> {
        let default_conf = PolicyConf::default();
        let mut filename = "".to_string();
        let mut policy_conf = match policy_file {
            Some(t) if t.len() > 0 => {
                filename = t.to_owned();
                let file = File::open(t)?;
                let mut read_conf: PolicyConf = serde_yaml::from_reader(file)?;
                read_conf.traced_syscalls = default_conf.traced_syscalls;
                read_conf
            },
            _ => {
                PolicyConf::default()
            }
        };

        if policy_conf.diff_to_default {
            let mut merged_extra_envs: HashMap<String, String> = HashMap::new();
            let mut merged_preserved_env_keys: HashSet<String> = HashSet::new();
            let mut merged_allowed_syscalls: HashSet<String> = HashSet::new();

            for env in default_conf.extra_envs {
                let split: Vec<_> = env.split('=').collect();
                merged_extra_envs.insert(split[0].to_owned(), split[1].to_owned());
            }
            for env in policy_conf.extra_envs {
                let split: Vec<_> = env.split('=').collect();
                merged_extra_envs.insert(split[0].to_owned(), split[1].to_owned());
            }
            policy_conf.extra_envs = Vec::new();
            for (key, value) in merged_extra_envs.into_iter() {
                policy_conf.extra_envs.push(format!("{}={}", key, value));
            }

            for key in default_conf.preserved_env_keys {
                merged_preserved_env_keys.insert(key);
            }
            for key in policy_conf.preserved_env_keys {
                merged_preserved_env_keys.insert(key);
            }
            policy_conf.preserved_env_keys = Vec::new();
            for key in merged_preserved_env_keys {
                policy_conf.preserved_env_keys.push(key);
            }

            for key in default_conf.allowed_syscalls {
                merged_allowed_syscalls.insert(key);
            }
            for key in policy_conf.allowed_syscalls {
                merged_allowed_syscalls.insert(key);
            }
            policy_conf.allowed_syscalls = Vec::new();
            for key in merged_allowed_syscalls {
                policy_conf.allowed_syscalls.push(key.to_string());
            }
        }

        Ok(
            FileBasedPolicy {
                file_name: filename.to_string(),
                conf: policy_conf,
            }
        )
    }
}

impl SandboxPolicy for FileBasedPolicy {
    fn check_path_op(&self, path: &str, op: PathOps, mode: i32) -> bool {
        let matchers = self.conf.whitelist_paths.get(&op).expect(&format!("Op {:?} does not exist", op));
        for matcher in matchers {
            let pattern = glob::Pattern::new(matcher).unwrap();
            if pattern.matches(path) {
                return true;
            }
        }
        return false;
    }

    fn get_exec_allowance(&self) -> i32 {
        let ret = self.conf.exec_allowance;
        return ret;
    }

    fn get_fork_allowance(&self) -> i32 {
        let ret = self.conf.fork_allowance;
        return ret;
    }

    fn get_max_child_procs(&self) -> i32 {
        let ret = self.conf.max_child_procs;
        return ret;
    }

    fn get_extra_envs(&self) -> Vec<String> {
        return self.conf.extra_envs.to_vec();
    }

    fn get_preserved_env_keys(&self) -> Vec<String> {
        return self.conf.preserved_env_keys.to_vec();
    }

    fn get_traced_syscalls(&self) -> Vec<String> {
        return self.conf.traced_syscalls.to_vec();
    }

    fn get_allowed_syscalls(&self) -> Vec<String> {
        return self.conf.allowed_syscalls.to_vec();
    }
}
